python/change-notes/2021-02-23-port-insecure-default-protocol.md

```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* Changed the query that detects insecure protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Modern versions of Python include fluent APIs that change default values after construction, so the query now reports results only in versions of Python where fluent APIs are not available.
```

Comment:
I don't understand this use of `fluent APIs`. Can you explain it in more detail? :blush:

Reply:
The appearance of "fluent APIs" here is my suggestion based on the summary for the PR. I may well have misunderstood the original description and so made a bad suggestion here.

Reply:
I accepted this suggestion because I found that it summarised the situation quite concisely. I guess I have to check if the term "fluent API" is perhaps not as wide-spread as I thought...

Reply:
Hm, there is actually a wikipedia entry on the term, which has a narrower scope, namely where you can chain the calls. That would be something like
```python
context = ssl.SSLContext(ssl.PROTOCOL_TLS).withOption(ssl.OP_NO_TLSv1)
```
rather than
```python
context = ssl.SSLContext(ssl.PROTOCOL_TLS)
context.options |= ssl.OP_NO_TLSv1
```
If people generally expect the former when reading "fluent API" and only the latter is actually provided, then this term might not be entirely appropriate.

Reply:
```suggestion
* Changed the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Python 3.4 introduced new ways to specify protocol version (through modifying the `options` field), which means that using the default value _can_ be secure (and if they're not, this should be reported by the `py/insecure-protocol` query). This query now only reports results only in versions of Python prior to 3.4.
```

Reply:
Thanks, this looks great apart from the duplicated "only" in the last sentence.